### PR TITLE
Date picker

### PR DESCRIPTION
**Pull request by spireken**

### CODE REVIEW

File: Applications/Spire/Source/Ui/CalendarDatePicker.cpp [hunk 1]
```diff
@@ -320,3 +337,91 @@ connection CalendarDatePicker::connect_submit_signal(
...
+          return true;
+        }
+      }
+    }
+  }
+  return QWidget::eventFilter(watched, event);
+}
+
+boost::optional<int> CalendarDatePicker::get_index(date day) const {
+  for(auto i = 0; i < m_calendar_view->get_list_model()->get_size(); ++i) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/CalendarDatePicker.cpp:366]: Avoid ```O(N)``` operations.

File: Applications/Spire/Source/Ui/CalendarDatePicker.cpp [hunk 2]
```diff
@@ -320,3 +337,91 @@ connection CalendarDatePicker::connect_submit_signal(
...
+    set_current_index({});
+  }
+}
+
+void CalendarDatePicker::on_current_month(date month) {
+  auto current_index = optional<int>();
+  auto list_has_focus = m_calendar_view->hasFocus() ||
+    m_calendar_view->isAncestorOf(focusWidget());
+  for(auto i = 0; i < m_calendar_view->get_list_model()->get_size(); ++i) {
+    auto minimum = m_model->get_minimum();
```
Comment (kranar) [inline Applications/Spire/Source/Ui/CalendarDatePicker.cpp:393]: How do ```minimum``` and ```maximum``` depend on ```i```?

File: Applications/Spire/Source/Ui/CalendarDatePicker.cpp [hunk 3]
```diff
@@ -320,3 +337,91 @@ connection CalendarDatePicker::connect_submit_signal(
...
+    auto maximum = m_model->get_maximum();
+    auto current = m_calendar_view->get_list_model()->get<date>(i);
+    if(minimum && maximum) {
+      m_calendar_view->get_list_item(i)->setDisabled(
+        current < *minimum || current > *maximum);
+    } else {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/CalendarDatePicker.cpp:399]: Factor this out.

File: Applications/Spire/Source/Ui/CalendarDatePicker.cpp [hunk 4]
```diff
@@ -320,3 +337,91 @@ connection CalendarDatePicker::connect_submit_signal(
...
+
+void CalendarDatePicker::on_current_month(date month) {
+  auto current_index = optional<int>();
+  auto list_has_focus = m_calendar_view->hasFocus() ||
+    m_calendar_view->isAncestorOf(focusWidget());
+  for(auto i = 0; i < m_calendar_view->get_list_model()->get_size(); ++i) {
+    auto minimum = m_model->get_minimum();
+    auto maximum = m_model->get_maximum();
+    auto current = m_calendar_view->get_list_model()->get<date>(i);
+    if(minimum && maximum) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/CalendarDatePicker.cpp:396]: I believe you can eliminate the test for ```minimum``` and ```maximum```. Review the following:

https://www.boost.org/doc/libs/1_77_0/libs/optional/doc/html/boost_optional/tutorial/relational_operators.html

A ```boost::none``` compares less than any other value.

File: Applications/Spire/Source/Ui/CalendarDatePicker.cpp [hunk 5]
```diff
@@ -320,3 +337,91 @@ connection CalendarDatePicker::connect_submit_signal(
...
+      m_calendar_view->get_list_item(i)->setDisabled(
+        current < *minimum || current > *maximum);
+    } else {
+      if(minimum) {
+        m_calendar_view->get_list_item(i)->setDisabled(current < *minimum);
+      } else if(maximum) {
+        m_calendar_view->get_list_item(i)->setDisabled(current > *maximum);
+      }
+    }
+    if(auto current_date = m_model->get_current(); current == current_date) {
```
Comment (kranar) [inline Applications/Spire/Source/Ui/CalendarDatePicker.cpp:406]: What is the purpose of ```current_date```?

File: Applications/Spire/Source/Ui/CalendarDatePicker.cpp [hunk 6]
```diff
@@ -307,8 +322,10 @@ CalendarDatePicker::CalendarDatePicker(
     set(BackgroundColor(QColor(0x4B23A0))).
     set(border(0, QColor(Qt::transparent))).
     set(TextColor(QColor(0xFFFFFF)));
-  set_style(*calendar_view, std::move(calendar_style));
-  layout->addWidget(calendar_view);
+  set_style(*m_calendar_view, std::move(calendar_style));
+  setFocusProxy(m_calendar_view);
+  setFixedWidth(scale_width(176));
```
Comment (shaobinspire) [inline Applications/Spire/Source/Ui/CalendarDatePicker.cpp:327]: I think the width can be derived from the width of m_calendar_view (168px) and the horizontal margins (4px for left and 4px for right).

**Pull request merged**